Packages/io.chainsafe.web3-unity/Editor/ABICSharpConverter.cs

```diff
@@ -0,0 +1,496 @@
+            return _instance;
+        }
+        set => _instance = value;
+    }
+
+    // UI rendering method
+        private void OnGUI()
+    {
+        var style = new GUIStyle(GUI.skin.label) { richText = true };
+
+        GUILayout.Space(10);
+        GUILayout.Label("<b><size=15>ChainSafe ABI to C# Contract Converter</size></b>", style);
+        GUILayout.Space(10);
+
+        _contractName = EditorGUILayout.TextField("Contract Name", _contractName);
+        _targetFolder = (DefaultAsset)EditorGUILayout.ObjectField("Select Folder", _targetFolder, typeof(DefaultAsset), false);
+
+        if (_targetFolder == null)
+        {
+            EditorGUILayout.HelpBox("No folder selected, please select a folder first", MessageType.Error);
+            return;
+        }
+
+        EditorGUI.BeginChangeCheck();
+        _abi = EditorGUILayout.TextField("ABI", _abi, EditorStyles.textArea, GUILayout.Height(200));
+        if (EditorGUI.EndChangeCheck())
+            _abiIsValid = IsValidAbi(_abi);
+
+        if (!_abiIsValid)
+        {
+            EditorGUILayout.HelpBox("Invalid ABI", MessageType.Error);
+            return;
+        }
+
+        if (string.IsNullOrEmpty(_contractName.Trim()))
+        {
+            EditorGUILayout.HelpBox("Contract Name cannot be empty", MessageType.Error);
+            return;
+        }
+
+        if (Regex.IsMatch(_contractName, @"[^a-zA-Z0-9_]|(?<=^)[0-9]"))
+        {
+            EditorGUILayout.HelpBox("Contract name is invalid", MessageType.Error);
+            return;
+        }
+
+        if (GUILayout.Button("Convert"))
+        {
+            ConvertAbiToCSharp();
+        }
+    }
+
+    [MenuItem("ChainSafe SDK/ABI to C# Contract Converter")]
+    public static void ShowWindow()
+    {
+        Instance = GetWindow<ABICSharpConverter>("ABI to C# Contract Converter");
+    }
+
+    // Method to handle ABI conversion
+    private void ConvertAbiToCSharp()
+    {
+        _contractABI = ABIDeserialiserFactory.DeserialiseContractABI(_abi);
+        var text = Resources.Load<TextAsset>("ABIContractClassTemplate").text;
+        var className = Regex.Replace(_contractName, @"[^a-zA-Z0-9_]|(?<=^)[0-9]", string.Empty);
+
+        text = text.Replace("{CLASS_NAME}", className);
+        var minifiedJson = JsonDocument.Parse(_abi).RootElement.GetRawText();
+        var escapedJson = minifiedJson.Replace("\"", "\\\"");
+        text = text.Replace("{CONTRACT_ABI}", escapedJson);
+        text = Regex.Replace(text, @"\s*\{CUSTOM_CLASSES\}", "\n\n" + ParseCustomClasses());
+        text = Regex.Replace(text, @"\s*\{EVENT_CLASSES\}", "\n\n" + ParseEventClasses());
+        text = Regex.Replace(text, @"\s*\{METHODS\}", "\n\n" + ParseMethods());
+        text = Regex.Replace(text, @"\s*\{EVENT_SUBSCRIPTION\}", "\n\n" + ParseEventSubscription());
+        text = Regex.Replace(text, @"\s*\{EVENT_UNSUBSCRIPTION\}", "\n\n" + ParseEventUnsubscription());
```

Comment:
I am quite shocked that you did everything using basic string operations and not some library. Wow.

Reply:
My secret obsession is actually to create code generations. Ever since we created a small compiler in college, I was amazed at how powerful that is :D